src/mlmc/quantity_concept.py

```diff
@@ -1,115 +1,430 @@
+        ts_type = TimeSeriesType(q_spec.times, field_type)
+        dict_types.append((q_spec.name, ts_type))
+    dict_type = DictType(dict_types)
+
+    return QuantityStorage(storage, dict_type)
+
+
+def remove_nan_samples(chunk):
+    """
+    Mask out samples that contain NaN in either fine or coarse part of the result
+    :param chunk: np.ndarray [M, chunk_size, 2]
+    :return: np.ndarray
+    """
+    # Fine and coarse moments mask
+    mask = np.any(np.isnan(chunk), axis=0)
+    m = ~mask.any(axis=1)
+    return chunk[..., m, :]
 
 
 def estimate_mean(quantity):
     """
-    Concept of the MLMC mean estimator.
-    TODO:
-    - test
-    - calculate variance estimates as well
-    :param quantity:
-    :return:
+    MLMC mean estimator.
+    The MLMC method is used to compute the mean estimate to the Quantity dependent on the collected samples.
+    The squared error of the estimate (the estimator variance) is estimated using the central limit theorem.
+    Data is processed by chunks, so that it also supports big data processing
+    :param quantity: Quantity
+    :return: QuantityMean which holds both mean and variance
     """
-    quantity_vec_size = quantity._size()
-    n_levels = 1
+    Quantity.samples.cache_clear()
+    quantity_vec_size = quantity.size()
     n_samples = None
     sums = None
+    sums_of_squares = None
     i_chunk = 0
-    while True:
-        chunk = quantity._samples(i_chunk)
-        if chunk is None:
-            break;
+    level_chunks_none = np.zeros(1)  # if ones than all level chunks are empty (None)
+
+    while not np.alltrue(level_chunks_none):
+        level_ids = quantity.level_ids()
         if i_chunk == 0:
             # initialization
-            n_levels = len(chunk)
+            n_levels = len(level_ids)
             n_samples = [0] * n_levels
-            sums = [np.zeros(chunk[0].shape[1]) for level in chunk]
-        for i_level, level_chunk in enumerate(chunk):
-            # level_chunk is Numpy Array with shape [M, chunk_size, 2]
-            n_samples[i_level] += len(level_chunk)
-            assert(level_chunk.shape[1] == quantity_vec_size)
-            sums[i_level] += np.sum(level_chunk[:, :, 0] - level_chunk[:, :, 1], axis=1)
+
+        level_chunks_none = np.zeros(n_levels)
+        for level_id in level_ids:
+            # Chunk of samples for given level id
+            chunk = quantity.samples(level_id, i_chunk)
+
+            if chunk is not None:
+                if level_id == 0:
+                    # Set variables for level sums and sums of powers
+                    if i_chunk == 0:
+                        sums = [np.zeros(chunk.shape[0]) for _ in range(n_levels)]
+                        sums_of_squares = [np.zeros(chunk.shape[0]) for _ in range(n_levels)]
+
+                    # Coarse result for level 0, there is issue for moments processing (not know about level)
+                    chunk[..., 1] = 0
+
+                chunk = remove_nan_samples(chunk)
```

Comment:
We should be able to report number of removed samples.

Reply:
separate issue created